controllers/openstackcontrolplane_controller.go

```diff
@@ -695,20 +591,454 @@ func (r *OpenStackControlPlaneReconciler) createOrUpdateOpenStackMACAddress(inst
+			err = common.WrapErrorForObject(cond.Message, instance, err)
+
+			return newVMs, err
+		}
+	}
+
+	return newVMs, nil
+}
+
+//
+// Create VMSets
+//
+func (r *OpenStackControlPlaneReconciler) createOrUpdateVMSets(
+	instance *ospdirectorv1beta1.OpenStackControlPlane,
+	cond *ospdirectorv1beta1.Condition,
+	deploymentSecret *corev1.Secret,
+) ([]*ospdirectorv1beta1.OpenStackVMSet, error) {
+	vmSets := []*ospdirectorv1beta1.OpenStackVMSet{}
+
+	for _, vmRole := range instance.Spec.VirtualMachineRoles {
+
+		//
+		// Create or update the vmSet CR object
+		//
+		vmSet := &ospdirectorv1beta1.OpenStackVMSet{
+			ObjectMeta: metav1.ObjectMeta{
+				// use the role name as the VM CR name
+				Name:      strings.ToLower(vmRole.RoleName),
+				Namespace: instance.Namespace,
+			},
+		}
+
+		op, err := controllerutil.CreateOrUpdate(context.TODO(), r.Client, vmSet, func() error {
+			vmSet.Spec.VMCount = vmRole.RoleCount
+			vmSet.Spec.Cores = vmRole.Cores
+			vmSet.Spec.Memory = vmRole.Memory
+			vmSet.Spec.DiskSize = vmRole.DiskSize
+			if instance.Spec.DomainName != "" {
+				vmSet.Spec.DomainName = instance.Spec.DomainName
+			}
+			vmSet.Spec.BootstrapDNS = instance.Spec.DNSServers
+			vmSet.Spec.DNSSearchDomains = instance.Spec.DNSSearchDomains
+			if vmRole.StorageClass != "" {
+				vmSet.Spec.StorageClass = vmRole.StorageClass
+			}
+			vmSet.Spec.BaseImageVolumeName = vmRole.DeepCopy().BaseImageVolumeName
+			vmSet.Spec.DeploymentSSHSecret = deploymentSecret.Name
+			vmSet.Spec.CtlplaneInterface = vmRole.CtlplaneInterface
+			vmSet.Spec.Networks = vmRole.Networks
+			vmSet.Spec.RoleName = vmRole.RoleName
+			vmSet.Spec.IsTripleoRole = vmRole.IsTripleoRole
+			if instance.Spec.PasswordSecret != "" {
+				vmSet.Spec.PasswordSecret = instance.Spec.PasswordSecret
+			}
+
+			err := controllerutil.SetControllerReference(instance, vmSet, r.Scheme)
+			if err != nil {
+				cond.Message = fmt.Sprintf("Error set controller reference for %s %s", vmSet.Kind, vmSet.Name)
+				cond.Reason = ospdirectorv1beta1.ConditionReason(ospdirectorv1beta1.CommonCondReasonControllerReferenceError)
+				cond.Type = ospdirectorv1beta1.ConditionType(ospdirectorv1beta1.CommonCondTypeError)
+				err = common.WrapErrorForObject(cond.Message, instance, err)
+
+				return err
+			}
+
+			return nil
+		})
+
+		if err != nil {
+			cond.Message = fmt.Sprintf("Failed to create or update %s %s ", vmSet.Kind, vmSet.Name)
+			cond.Reason = ospdirectorv1beta1.ConditionReason(ospdirectorv1beta1.VMSetCondReasonError)
+			cond.Type = ospdirectorv1beta1.ConditionType(ospdirectorv1beta1.CommonCondTypeError)
+			err = common.WrapErrorForObject(cond.Message, instance, err)
+
+			return vmSets, err
+		}
+		if op != controllerutil.OperationResultNone {
+			cond.Message = fmt.Sprintf("%s %s operation: %s", vmSet.Kind, vmSet.Name, string(op))
+			cond.Reason = ospdirectorv1beta1.ConditionReason(ospdirectorv1beta1.VMSetCondReasonProvisioned)
```

Comment:
Isn't the VMSet just created/updated here, and not necessarily provisioned yet?

Reply:
yes correct, will update that here